File: Python/P1/P1_E1.py
# Aprendizaje Automático 2023 - 2024

# Práctica 1 - Ejercicio 1

# Julia López Augusto
# Gonalo Vega Pérez

import numpy as np
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

if __name__ == '__main__':

    # Datos Iniciales
    x1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    x2 = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
    X = np.column_stack((x1, x2))
    Y = np.array([1.56, 1.95, 2.44, 3.05, 3.81, 4.77, 5.96, 7.45, 9.31, 11.64])


    # Crear Modelo y entrenarlo
    mdl = LinearRegression().fit(X,Y)

    # Resultados (Pesos)
    coef = mdl.coef_
    intercept = mdl.intercept_

    print("Coef:" + str(coef))
    print("Intercept:" + str(intercept))

    # Utilizar modelo entrenado para predecir resultados
    Ye = mdl.predict(X)

    # Representar graficamente
    fig = plt.figure
    ax = plt.axes(projection='3d')

    # Mostrar datos iniciales
    ax.plot3D(x1, x2, Y) 
    # Mostrar predicción
    ax.plot3D(x1, x2, Ye, 'red') 

    # Nombrar Ejes
    ax.set_xlabel('x1', labelpad=20)
    ax.set_ylabel('x2', labelpad=20)
    ax.set_zlabel('y', labelpad=20)

    plt.show()

File: Matlab/P1/P1_E1.m
% Ejercicio 1. Regresión Lineal
% Julia López
% Gonzalo Vega
% AA - 2023

clc
clear

x1 = [1,2,3,4,5,6,7,8,9,10];
x2 = [1,1,1,1,1,1,1,1,1,1];
X = [x1;x2].';

Y = [0, 0.69, 1.1, 1.39, 1.61, 1.79, 1.95, 2.08, 2.2, 2.3];

% --- 1.1 ---
% Crear modelo de regresión lineal
mdl = fitlm(X,Y)

%% 

% --- 1.2 ---
% Calcular predicciones
ye = predict(mdl, X);

% Mostrar datos iniciales
scatter3(x1, x2 ,Y, 'b*');
hold on 
% Mostrar predicción
plot3(x1,x2, ye, 'r');
hold off

%% 

% --- 1.4 ---
% Obtener función de coste (J)
syms w1 w2;
w0 = 0.24;
%w1 = 0.23;
%w2 = 0;
W = [w1,w2];
n = 10;  % número de muestras
total = 0;
for c = 1:2
    for i = 1:n
        h = w0 + W(c)*X(i, c);
        total = total + (h - Y(i))^2;
    end
end

% Función de coste
J = (1/(2*n)) * total;

fmesh(J)